### ISSUE TITLE
"too many range table entries" on nested filtering in the presence of Access Policies

### ISSUE THREAD

**Issue by jackfischer** [edgedb/edgedb#4564]
- EdgeDB Version: 3.0-dev.7000+58a0109 
- OS Version: macOS 12.6

Steps to Reproduce:

1. Perform a highly nested select/filter with access policies enabled (example below). We are seeing this on other queries as well. No inserted data required.
`edgedb error: InvalidValueError: too many range table entries`

To restore normal edgedb behavior, disable access policies via `configure session set apply_access_policies := false`

QB

```
e.select(e.StrategyNode, node => ({
    tiptapId: true,
    metrics: metric => ({
      predictions: e.select(metric.predictingQuestion.responses, response => ({
        content: true,
        createdDatetime: true,
        limit: 1,
        order_by: {
          expression: response.createdDatetime,
          direction: e.DESC,
          empty: EMPTY_LAST,
        },
        target: e.select(metric.metricTargets, target => ({
          id: true,
          targetValue: true,
          filter: e.op(target.strategyNode, "=", e.select(e.StrategyNode, node => ({
            filter: e.op(node.tiptapId, "=", node.tiptapId),
          }))),
        })),
        tiptapId: node.tiptapId,
      })),
    }),
  }));
```

Generated EdgeQL

```
WITH
  __scope_0_defaultStrategyNode := DETACHED default::StrategyNode
SELECT __scope_0_defaultStrategyNode {
  tiptapId,
  metrics := (
    WITH
      __scope_1_defaultMetric := __scope_0_defaultStrategyNode.metrics
    SELECT __scope_1_defaultMetric {
      multi predictions := (
        WITH
          __scope_2_defaultSignal := __scope_1_defaultMetric.predictingQuestion.responses
        SELECT __scope_2_defaultSignal {
          content,
          createdDatetime,
          multi target := (
            WITH
              __scope_3_defaultMetricTarget := __scope_1_defaultMetric.metricTargets
            SELECT __scope_3_defaultMetricTarget {
              id,
              targetValue
            }
            FILTER (__scope_3_defaultMetricTarget.strategyNode = (WITH
              __scope_4_defaultStrategyNode := DETACHED default::StrategyNode
            SELECT __scope_4_defaultStrategyNode {
              id
            }
            FILTER (__scope_4_defaultStrategyNode.tiptapId = __scope_0_defaultStrategyNode.tiptapId)))
          ),
          single tiptapId := __scope_0_defaultStrategyNode.tiptapId
        }
        ORDER BY __scope_2_defaultSignal.createdDatetime DESC EMPTY LAST
        LIMIT 1
      )
    }
  )
}
```

Schema:

```
module default {
    global current_user -> uuid;

    type ActivitySource {
        required link strategyNode -> StrategyNode;
        property sourceName -> str;
        property sourceUrl -> str;
        property timeLastPulled -> datetime {
            default := datetime_current();
        }

        constraint exclusive on ((.sourceUrl, .strategyNode));
    }

    type DashboardConfig {
        required link strategyNode -> StrategyNode;
        required property dashboardSource -> str;
        multi link metrics -> Metric;
    }

    type Document extending SignalSubject {
        property unpublishedVersion -> json; # May eventually be some collaborative editing supportive representation, just a serialized document for now. #TODO make required
        link mostRecentPublished := (
            select .<parentDocument[is DocumentVersion] order by .createdDatetime desc limit 1
        );
        multi link history := (
            select .<parentDocument[is DocumentVersion] order by .createdDatetime desc
        );
        required property ontology -> array<str>;
        required link author -> User;
        multi link viewers -> User;
        multi link editors -> User;
        multi link owners -> User;

        constraint exclusive on (.author);

        access policy author_has_full_access
            allow all
            using (global current_user ?= .author.id);

        access policy owner_has_full_access
            allow all
            using ((global current_user in .owners.id) ?? false);

        access policy allow_select_or_update_for_editor
            allow select, update, update read, update write
            using ((global current_user in .editors.id) ?? false);

        access policy allow_select_for_viewer
            allow select
            using ((global current_user in .viewers.id) ?? false);

        access policy allow_insert
            allow insert
            using (true);
    }

    type DocumentVersion {
        required link parentDocument -> Document;
        required property content -> json;
        required property ontology -> array<str>;
        required property createdDatetime -> datetime {
            default := datetime_current();
            readonly := true;
        }
        index on ((.parentDocument, .createdDatetime));

        access policy author_has_full_access
            allow all
            using (global current_user ?= .parentDocument.author.id);

        access policy owner_has_full_access
            allow all
            using ((global current_user in .parentDocument.owners.id) ?? false);

        access policy allow_select_or_update_for_editor
            allow select, update, update read, update write
            using ((global current_user in .parentDocument.editors.id) ?? false);

        access policy allow_select_for_viewer
            allow select
            using ((global current_user in .parentDocument.viewers.id) ?? false);

        access policy allow_insert
            allow insert
            using (true);
    }

    type Event extending SignalSubject {
        required property eventDatetime -> datetime {
            default := datetime_current();
        }
        required property name -> str;
        property type_ -> str;
        required link author -> User;
        multi link strategyNodes -> StrategyNode;
    }


    type Metric extending SignalSubject {
        property name -> str {
            annotation description := "Optional, likely to be exactly the same as the SN and someday maybe a computed property, but may have an idiosyncratic name"
        }
        property type_ -> str;
        multi link strategyNodes -> StrategyNode {
            property target_ -> float64; # Unused. Will be used instead of the target object when link properties are better supported
        }
        required property sourceMetadata -> json;
        multi link metricTargets := .<metric[is MetricTarget];
        link generatingQuestion := .<generatedMetric[is Question];
        link predictingQuestion := .<predictedMetric[is Question];
    }

    type MetricTarget extending SignalSubject {
        property name -> str;
        property targetValue -> float64;
        required link author -> User;
        required link strategyNode -> StrategyNode;
        required link metric -> Metric;
        property score -> float64;
        property scoringFunction -> json;
        property targetCompletionDate -> datetime;
    }

    abstract type Principal extending SignalSubject {
        required property name -> str;
        link parentTeam -> Team;
    }

    type Question {
        required property type_ -> str;
        required property prompt -> str;
        link predictedMetric -> Metric;
        link generatedMetric -> Metric;
        multi link positions := .<question[is QuestionPosition];
        multi link responses := .<question[is Signal];
    }

    type QuestionPosition {
        required property position -> int32;
        required link question -> Question;
        required link survey -> Survey;
    }


    type Signal extending SignalSubject { # Extends SignalSubject to enable conversation-style signal
        required property content -> json;
        required property type_ -> str;
        link subject -> SignalSubject;
        link strategyNode ->StrategyNode;
        link survey -> Survey;
        link question -> Question;

        required link author -> User; # TODO make required when we have a auth and users live
        multi link recipients -> Principal; # TODO new recipients permissions model

        required property createdDatetime -> datetime {
            default := datetime_current();
            #readonly := true;
        }
        property modifiedDatetime -> datetime;
        constraint exclusive on ((.subject, .author, .createdDatetime, .type_))
    }

    abstract type SignalSubject {
        multi link signals := .<subject[is Signal];
    }

    type StrategyNode extending SignalSubject {
        link parentNode -> StrategyNode;
        multi link childNodes := .<parentNode[is StrategyNode];

        required link documentVersion -> DocumentVersion; # Latest available document version for this node.

        required property nodeType -> str; # This may eventually reference a node type as data with ontology customization
        required property title -> str;
        required property tiptapId -> str {
            constraint exclusive;
        };
        property metadata -> json;

        multi link metrics := .<strategyNodes[is Metric];

        multi link responsibleParties -> User;
        multi link defaultSignalViewers -> Principal;
        multi link subscribers -> User;
        property status -> str;
        link team -> Team;
        access policy author_has_full_access
            allow all
            using (global current_user ?= .documentVersion.parentDocument.author.id);

        access policy owner_has_full_access
            allow all
            using ((global current_user in .documentVersion.parentDocument.owners.id) ?? false);

        access policy allow_select_or_update_for_editor
            allow select, update, update read, update write
            using ((global current_user in .documentVersion.parentDocument.editors.id) ?? false);

        access policy allow_select_for_viewer
            allow select
            using ((global current_user in .documentVersion.parentDocument.viewers.id) ?? false);

        access policy allow_insert
            allow insert
            using (true);

        index on (.tiptapId);

    }

    type StrategyNodeEdit extending SignalSubject {
        property fromStatus -> str;
        property toStatus -> str;
        property fromTitle -> str;
        property toTitle -> str;
        property fromType -> str;
        property toType -> str;
        property oldNode -> json;
        property newNode -> json;
        required property editedFields -> array<str>;
        property reason -> str;
        multi link motivatingSignal -> Signal;
        required link author -> User;
        required property createdDatetime -> datetime {
            default := datetime_current();
            #readonly := true;
        }
        # access policy author_has_full_access
        #     allow all
        #     using (global current_user ?= .documentVersion.parentDocument.author.id);

        # access policy owner_has_full_access
        #     allow all
        #     using ((global current_user in .documentVersion.parentDocument.owners.id) ?? false);

        # access policy allow_select_or_update_for_editor
        #     allow select, update, update read, update write
        #     using ((global current_user in .documentVersion.parentDocument.editors.id) ?? false);

        # access policy allow_select_for_viewer
        #     allow select
        #     using ((global current_user in .documentVersion.parentDocument.viewers.id) ?? false);

        # access policy allow_insert
        #     allow insert
        #     using (true);
    }

    type Survey {
        required property name -> str;
        required property isRecurring -> str;
        property reason -> str;
        multi link recipients -> User;
        required multi link questions -> Question {
            property position -> int64; # Unused. Will be used instead of position object when link properties are better supported
        }
        required property subjectTypes -> array<str>;
        required property subjects -> array<str>; # TO DO - should probably be a multi link to strategy node but want to require the prop for now since we are not yet populating strategy node
        required link strategyNode -> StrategyNode;
        required link author -> User;
    }


    type Team extending Principal {
        required multi link leads -> User;  # TODO should probably be a property on the link
        required multi link members -> User;
        multi link subTeams -> Team;
        multi link subscribers -> User;
    }

    type User extending Principal {
        required property email -> str;
        required property provider -> str;
        property given_name -> str;
        property family_name -> str;
        property picture -> str;
        property email_verified -> bool;
        property locale -> str;
        property provider_id -> str;
        multi link canView := .<viewers[is Document];
        multi link canEdit := .<editors[is Document];
        multi link authored := .<author[is Document];

        constraint exclusive on ((.email, .provider));
    }

}
```

**Comment by msullivan** [comment]
The cause of this here is that current, access policies on other types are respected while evaluating access policies. So during the evaluation of `StrategyNode`, the access policies for both `DocumentVersion` and `Document` are evaluated, and while evaluating the policies for `DocumentVersion`, the access policies for `Document` are evaluated. Our access policy filters are compiled into CTEs in the generated SQL, but Postgres inlines those, causing an enormous blowup--there will be over 100 evaluations of `Document` policies.

The decision to respect other types' access policies during the evaluation of policies was made for a few reasons, including that it matches the design of row level security in postgres and that it fit with some of our future plans for Role Based Access Control, where eventually it will be possible for a not fully trusted user to have write access to parts of the schema but not all of it.

But we now think that this decision, in the context of EdgeDB, was probably not the optimal one. Your use case, where the policies for many objects is driven by fields on a target, is a powerful one, and I don't think it is supported well with the current behavior.

Our next point release, 2.6, will include a mechanism for schemas to opt-in to what should eventually become the default behavior, in which access policies themselves have unfettered access to the database.

**Comment by jackfischer** [comment]
Many thanks for the detailed information here. Looking forward to the opt out setting in 2.6.

**Comment by msullivan** [comment]
Alright, we've released 2.6 with nonrecursive_policies. I'm going to close this.
